File: dynamic-programming/distinct-subsequences-ii.java
class Solution {
    public int distinctSubseqII(String s) {
        int n = s.length();
        int[] dp = new int[n + 1];
        dp[0] = 1;
        int Mod = 1000000007;
        HashMap<Character, Integer> mp = new HashMap<>();

        for (int i = 1; i <= n; i++) {
            char ch = s.charAt(i - 1);
            dp[i] = (dp[i - 1] * 2) % Mod;
            if (mp.containsKey(ch)) {
                int last_ind = mp.get(ch);
                dp[i] = (dp[i] - dp[last_ind - 1] + Mod) % Mod;
            }
            mp.put(ch, i);
        }
        dp[n] = (dp[n] - 1 + Mod) % Mod;
        return dp[n];
    }
}

File: dynamic-programming/distinct-subsequences-ii.cpp
class Solution {
public:
    int Mod = 1e9 + 7;
    int n;
    int solve(int i, int prev, string& s, vector<vector<int>>& dp) {
        if (i >= n) {
            return 0;
        }
        if (prev != -1 && dp[i][prev] != -1) {
            return dp[i][prev];
        }
        int ans = 0;
        // if same as prev move forward
        if (prev != -1 && s[i] == s[prev]) {
            ans = (ans + solve(i + 1, prev, s, dp)) % Mod;
        } else {
            // if not either take it or not take it
            ans = (ans + solve(i + 1, prev, s, dp)) % Mod;
            ans = (ans + 1 + solve(i + 1, i, s, dp)) % Mod;
        }
        if (prev != -1) {
            dp[i][prev] = ans;
        }
        return ans;
    }

    int distinctSubseqII(string s) {
        n = s.length();
        vector<vector<int>> dp(n, vector<int>(n, -1));
        return solve(0, -1, s, dp);
    }
};